src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,539 @@
+            .vk
+            .domain
+            .coeff_to_extended(permuted_table_poly.clone(), Rotation::default());
+        let permuted_table_blind = Blind(C::Scalar::random());
+        let permuted_table_commitment = params
+            .commit_lagrange(&permuted_table_value, permuted_table_blind)
+            .to_affine();
+
+        let permuted = Permuted {
+            permuted_input_value,
+            permuted_input_poly,
+            permuted_input_coset,
+            permuted_input_inv_coset,
+            permuted_input_blind,
+            permuted_input_commitment,
+            permuted_table_value,
+            permuted_table_poly,
+            permuted_table_coset,
+            permuted_table_blind,
+            permuted_table_commitment,
+        };
+
+        self.permuted = Some(permuted.clone());
+        permuted
+    }
+
+    fn permute_column_pair(
+        input_value: &Polynomial<C::Scalar, LagrangeCoeff>,
+        table_value: &Polynomial<C::Scalar, LagrangeCoeff>,
+    ) -> (
+        Polynomial<C::Scalar, LagrangeCoeff>,
+        Polynomial<C::Scalar, LagrangeCoeff>,
+    ) {
+        let mut input_coeffs = input_value.get_values().to_vec();
+        let table_coeffs = table_value.get_values().to_vec();
+
+        // Sort input lookup column values
+        input_coeffs.sort();
+        let permuted_input_value = Polynomial::new(input_coeffs.to_vec());
+
+        // A BTreeMap of each unique element in the table column and its count
+        let mut leftover_table_map: BTreeMap<C::Scalar, u32> =
+            table_coeffs.iter().fold(BTreeMap::new(), |mut acc, coeff| {
+                *acc.entry(*coeff).or_insert(0) += 1;
+                acc
+            });
+        let mut repeated_input_rows = vec![];
+        let mut permuted_table_coeffs = vec![C::Scalar::zero(); table_coeffs.len()];
+
+        for row in 0..input_coeffs.len() {
+            let input_value = input_coeffs[row];
+
+            // If this is the first occurence of `input_value` in the input column
+            if row == 0 || input_value != input_coeffs[row - 1] {
+                permuted_table_coeffs[row] = input_value;
+                // Remove one instance of input_value from leftover_table_map
+                if let Some(count) = leftover_table_map.get_mut(&input_value) {
+                    assert!(*count > 0);
+                    *count -= 1;
+                } else {
+                    // Panic if input_value not found
+                    panic!("Input value not found in table.");
+                }
+            // If input value is repeated
+            } else {
+                repeated_input_rows.push(row);
+            }
+        }
+
+        // Populate permuted table at unfilled rows with leftover table elements
+        for (coeff, count) in leftover_table_map.iter() {
+            for _ in 0..*count {
+                permuted_table_coeffs[repeated_input_rows.remove(0) as usize] = *coeff;
```

Comment:
```suggestion
                permuted_table_coeffs[repeated_input_rows.pop() as usize] = *coeff;
```
[remove(0) is O(length)](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove), which would make the overall complexity quadratic. Since the order doesn't matter, we can use pop() which is O(1).